### PR TITLE
Fix race condition when creating build dir

### PR DESCRIPTION
**Pull request by Stonesjtu**
The original `check-and-act` style can raise `FileExistsError` when multiple processes are jit-compiling the extension on the same node.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@Stonesjtu you will need to find someone else to land this PR as I am out of country for three weeks

**Comment by mrshenli** [comment]
Looks like this PR breaks master. Is fix on the way or should we revert? 

https://app.circleci.com/jobs/github/pytorch/pytorch/4116989
https://app.circleci.com/jobs/github/pytorch/pytorch/4117323

example failure:

```
Jan 03 16:38:04 ======================================================================
Jan 03 16:38:04 ERROR [0.547s]: test_warning (__main__.TestCppExtension)
Jan 03 16:38:04 ----------------------------------------------------------------------
Jan 03 16:38:04 Traceback (most recent call last):
Jan 03 16:38:04   File "test_cpp_extensions.py", line 806, in test_warning
Jan 03 16:38:04     with_pytorch_error_handling=True)
Jan 03 16:38:04   File "/opt/python/2.7/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 759, in load_inline
Jan 03 16:38:04     build_directory = build_directory or _get_build_directory(name, verbose)
Jan 03 16:38:04   File "/opt/python/2.7/lib/python2.7/site-packages/torch/utils/cpp_extension.py", line 1055, in _get_build_directory
Jan 03 16:38:04     except FileExistsError:
Jan 03 16:38:04 NameError: global name 'FileExistsError' is not defined
```

cc @ezyang

**Comment by ezyang** [comment]
I was negligent in checking build status before landing. It looks like this diff needs a little more juice for 2.7 compat, or will have to wait until we drop 2.7 support on master (stalled due to some internal projects that haven't gotten off python 2)

**Comment by Stonesjtu** [comment]
@ezyang I can add a python2/3 compatible statement for that, but is there a road-map to drop python2 support?

**Comment by ezyang** [comment]
I think @suo would know better, c.f. #31164. They're going to start sending nastygrams to the service maintainers in February; hopefully we can get it fixed before then...

**Comment by ezyang** [comment]
Also, if you could reopen this PR and cc me to it, that would be much appreciated

**Comment by suo** [comment]
@Stonesjtu Unfortunately we do indeed need to support py2 for a few more months. The official drop-dead date for us is ~April.

**Comment by salexspb** [comment]
@suo , @ezyang , should this be re-merged again now? :)

**Comment by ezyang** [comment]
Yup, Py2 is dead. If you can just resubmit the pr I can get the ball rolling

**Comment by Stonesjtu** [comment]
So should we rebase first or just reopen it?

**Comment by ezyang** [comment]
Opening a new PR will be easiest for us. Add me as reviewer

**Pull request closed without merging**
